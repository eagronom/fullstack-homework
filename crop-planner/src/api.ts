import { Crop, Field, HumusBalanceAPIRespone } from "./types";

const SOIL_SERVICE_URL = "http://localhost:3000";

export const fetchFields = async (): Promise<Array<Field>> =>
  await fetch(`${SOIL_SERVICE_URL}/fields`).then((response) => response.json());

export const fetchCrops = async (): Promise<Array<Crop>> =>
  await fetch(`${SOIL_SERVICE_URL}/crops`).then((response) => response.json());

export const fetchHumusBalance = async (
  field: Field
): Promise<HumusBalanceAPIRespone> => {
  const params = `field_id=${field.id}&${field.crops
    .map((crop) => `years[${crop.year}]=${crop.crop?.value}`)
    .join("&")}`;
  return await fetch(
    `${SOIL_SERVICE_URL}/humus-balance/?${encodeURI(params)}`
  ).then((response) => response.json());
};
