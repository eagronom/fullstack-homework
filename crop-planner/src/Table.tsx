import { PureComponent } from 'react'
import {sortBy, find } from 'lodash'

import CropSelect from './CropSelect'
import {Crop, Field, HumusBalance, SeasonalCrop} from './types'
import { fetchCrops, fetchFields, fetchHumusBalance, calculateHumusBalance } from './api'
import buildNewFieldsState from './buildNewFieldsState'
import buildNewHumusState from "./buildNewHumusState";

type Props = {}

type State = {
  allCrops: Array<Crop>,
  fields: Array<Field>,
  humusBalance: Array<HumusBalance>
}

export default class Table extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      allCrops: [],
      fields: [],
      humusBalance: []
    }
  }

  componentDidMount = async () =>
    this.setState({
      fields: await fetchFields(),
      allCrops: await fetchCrops(),
      humusBalance: await fetchHumusBalance()
    })

  render = () =>
    <div className="table">
      <div className="table__row table__row--header">
        <div className="table__cell">Field name</div>
        <div className="table__cell table__cell--right">Field area (ha)</div>
        <div className="table__cell table__cell--center">2020 crop</div>
        <div className="table__cell table__cell--center">2021 crop</div>
        <div className="table__cell table__cell--center">2022 crop</div>
        <div className="table__cell table__cell--center">2023 crop</div>
        <div className="table__cell table__cell--center">2024 crop</div>
        <div className="table__cell table__cell--right">Humus balance</div>
      </div>

      {sortBy(this.state.fields, field => field.name).map(field => this.renderFieldRow(field))}
    </div>

  renderFieldRow = (field: Field) =>
    <div className="table__row" key={field.id}>
      <div className="table__cell">{field.name}</div>
      <div className="table__cell table__cell--right">{field.area}</div>

      {sortBy(field.crops, crop => crop.year).map(seasonalCrop => this.renderCropCell(field, seasonalCrop))}

      <div className="table__cell table__cell--right">{find(this.state.humusBalance, humusBalance => field.id === humusBalance.field_id)?.humus_balance ?? 'Not Calculated'}</div>
    </div>

  renderCropCell = (field: Field, seasonalCrop: SeasonalCrop) =>
    <div className="table__cell table__cell--center table__cell--with-select">
      <CropSelect
        selectedCrop={seasonalCrop.crop}
        allCrops={this.state.allCrops}
        onChange={newCrop => this.changeFieldCrop(newCrop, field.id, seasonalCrop.year)}
      />
    </div>

  changeFieldCrop = (newCrop: Crop | null, fieldId: number, cropYear: number) => {
      let newFieldsState = buildNewFieldsState(this.state.fields, newCrop, fieldId, cropYear)
      this.setState(newFieldsState)
      const changedField = find(newFieldsState.fields, field => field.id === fieldId)!

      let newBalance = calculateHumusBalance(changedField)
      newBalance.then((humusBalance) => {
          this.setState(buildNewHumusState(this.state.humusBalance, humusBalance))
      })


      return null
  }

}
