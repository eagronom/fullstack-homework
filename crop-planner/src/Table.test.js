import { shallow } from "enzyme";
import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Table from "./Table";
import { fetchCrops, fetchFields, fetchHumusBalance } from "./api";

jest.mock("./api");

describe("<Table />", () => {
  const crops = [{ value: 1, label: "Crop 1", humus_delta: 1 }];
  const fields = [
    {
      id: 1,
      name: "MÃ¤eotsa",
      area: 0.93,
      crops: [
        { year: 2020, crop: { value: 1, label: "Crop 1", humus_delta: 1 } },
        { year: 2021, crop: { value: 2, label: "Crop 2", humus_delta: 1 } },
        { year: 2022, crop: { value: 3, label: "Crop 3", humus_delta: 1 } },
        { year: 2023, crop: { value: 4, label: "Crop 4", humus_delta: 1 } },
        { year: 2024, crop: { value: 5, label: "Crop 5", humus_delta: 1 } },
      ],
    },
  ];

  const humusBalance = {
    field_id: 1,
    humus_balance: -4.3,
    previous_humus_balance: -3,
  };

  beforeEach(() => {
    fetchCrops.mockReturnValue(crops);
    fetchFields.mockReturnValue(fields);
    fetchHumusBalance.mockReturnValue({});
  });

  it("renders with empty data", async () => {
    expect(await shallow(<Table />)).toMatchSnapshot();
  });

  it("loads data when component mounted", async () => {
    expect(
      await shallow(<Table />)
        .instance()
        .componentDidMount()
    ).toMatchSnapshot();
  });

  it("loads data when component mounted with humus balance information", async () => {
    fetchHumusBalance.mockReturnValue(humusBalance);
    expect(
      await shallow(<Table />)
        .instance()
        .componentDidMount()
    ).toMatchSnapshot();
  });

  it("shows new humus balance value after select different crop", async () => {
    fetchCrops.mockReturnValue([
      ...crops,
      { value: 6, label: "Crop 6", humus_delta: 1 },
    ]);

    fetchHumusBalance
      .mockReturnValueOnce(humusBalance)
      .mockReturnValueOnce({ ...humusBalance, humus_balance: 100.0 });

    render(<Table />);

    const cropDropDown = await screen.findByText("Crop 1");
    userEvent.click(cropDropDown);

    const newCropOption = await screen.findByText("Crop 6");
    userEvent.click(newCropOption);
    expect(await screen.findByText(/100\.00/i)).toBeTruthy();
  });

  it("shows humus balance as good", async () => {
    fetchCrops.mockReturnValue([
      ...crops,
      { value: 6, label: "Crop 6", humus_delta: 1 },
    ]);

    const goodHumusBalance = 100.0;

    fetchHumusBalance
      .mockReturnValueOnce(humusBalance)
      .mockReturnValueOnce({
        ...humusBalance,
        humus_balance: goodHumusBalance,
      });

    render(<Table />);

    const cropDropDown = await screen.findByText("Crop 1");
    userEvent.click(cropDropDown);

    const newCropOption = await screen.findByText("Crop 6");
    userEvent.click(newCropOption);
    const humusBalanceText = await screen.findByText(
      goodHumusBalance.toFixed(2)
    );

    expect(humusBalanceText).toMatchSnapshot();
  });

  it("shows humus balance as bad balance", async () => {
    fetchCrops.mockReturnValue([
      ...crops,
      { value: 6, label: "Crop 6", humus_delta: 1 },
    ]);
    const badHumusBalance = humusBalance.previous_humus_balance - 1;
    fetchHumusBalance.mockReturnValueOnce(humusBalance).mockReturnValueOnce({
      ...humusBalance,
      humus_balance: badHumusBalance,
    });

    render(<Table />);

    const cropDropDown = await screen.findByText("Crop 1");
    userEvent.click(cropDropDown);

    const newCropOption = await screen.findByText("Crop 6");
    userEvent.click(newCropOption);
    const humusBalanceText = await screen.findByText(
      badHumusBalance.toFixed(2)
    );

    expect(humusBalanceText).toMatchSnapshot();
  });
});
