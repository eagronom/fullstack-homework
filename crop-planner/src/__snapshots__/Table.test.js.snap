// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> loads data when component mounted 1`] = `
<div
  className="table"
>
  <div
    className="table__row table__row--header"
  >
    <div
      className="table__cell"
    >
      Field name
    </div>
    <div
      className="table__cell table__cell--right"
    >
      Field area (ha)
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2020 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2021 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2022 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2023 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2024 crop
    </div>
    <div
      className="table__cell table__cell--right"
    >
      Humus balance
    </div>
  </div>
  <div
    className="table__row"
    key="1"
  >
    <div
      className="table__cell"
    >
      Mäeotsa
    </div>
    <div
      className="table__cell table__cell--right"
    >
      0.93
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 1",
            "value": 1,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 2",
            "value": 2,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 3",
            "value": 3,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 4",
            "value": 4,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 5",
            "value": 5,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--right"
    >
      --
    </div>
  </div>
</div>
`;

exports[`<Table /> loads data when component mounted with humus balance information 1`] = `
<div
  className="table"
>
  <div
    className="table__row table__row--header"
  >
    <div
      className="table__cell"
    >
      Field name
    </div>
    <div
      className="table__cell table__cell--right"
    >
      Field area (ha)
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2020 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2021 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2022 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2023 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2024 crop
    </div>
    <div
      className="table__cell table__cell--right"
    >
      Humus balance
    </div>
  </div>
  <div
    className="table__row"
    key="1"
  >
    <div
      className="table__cell"
    >
      Mäeotsa
    </div>
    <div
      className="table__cell table__cell--right"
    >
      0.93
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 1",
            "value": 1,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 2",
            "value": 2,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 3",
            "value": 3,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 4",
            "value": 4,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--center table__cell--with-select"
    >
      <CropSelect
        allCrops={
          Array [
            Object {
              "humus_delta": 1,
              "label": "Crop 1",
              "value": 1,
            },
          ]
        }
        onChange={[Function]}
        selectedCrop={
          Object {
            "humus_delta": 1,
            "label": "Crop 5",
            "value": 5,
          }
        }
      />
    </div>
    <div
      className="table__cell table__cell--right table__cell--bad-balance"
    >
      -4.30
    </div>
  </div>
</div>
`;

exports[`<Table /> renders with empty data 1`] = `
<div
  className="table"
>
  <div
    className="table__row table__row--header"
  >
    <div
      className="table__cell"
    >
      Field name
    </div>
    <div
      className="table__cell table__cell--right"
    >
      Field area (ha)
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2020 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2021 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2022 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2023 crop
    </div>
    <div
      className="table__cell table__cell--center"
    >
      2024 crop
    </div>
    <div
      className="table__cell table__cell--right"
    >
      Humus balance
    </div>
  </div>
</div>
`;
